File: apps/studio/authoring/host/studio_host/runtime.py
from __future__ import annotations

from typing import Any

from inpainter.errors import CoreError
from inpainter.lua_runtime import transition

from studio_host.contract import EFFECTS, SCRIPTS
from studio_host.paths import scripts_dir
from studio_host.schema import apply_defaults

MAX_DELEGATE_DEPTH = 8


def initialize() -> dict[str, Any]:
    return apply_defaults("session", {})


def run_script(script: str, state: dict[str, Any], event: dict[str, Any], depth: int = 0) -> dict[str, Any]:
    if script not in SCRIPTS:
        raise CoreError(f"unknown lua script: {script}")
    if depth > MAX_DELEGATE_DEPTH:
        raise CoreError(f"lua delegate depth exceeded: {script}")

    def delegate(name: str, child_state: Any, child_event: Any) -> dict[str, Any]:
        next_state = child_state if isinstance(child_state, dict) else {}
        next_event = child_event if isinstance(child_event, dict) else {}
        return run_script(str(name), next_state, next_event, depth + 1)

    result = transition(scripts_dir(), script, state, event, helpers={"delegate": delegate})
    return normalize(result)


def normalize(result: dict[str, Any]) -> dict[str, Any]:
    state = result.get("state")
    if not isinstance(state, dict):
        raise CoreError("Client Lua must return session state")
    effects = _as_list(result.get("effects"))
    for effect in effects:
        if not isinstance(effect, dict) or effect.get("type") not in EFFECTS:
            raise CoreError(f"unsupported client effect: {effect}")
        if effect["type"] in ("ui.feed.show", "operation.invoke"):
            effect["messages"] = _as_message_list(effect.get("messages"))
        if effect["type"] == "ui.header":
            effect["actions"] = _as_list(effect.get("actions"))
    feed = state.get("feed")
    if not isinstance(feed, dict):
        feed = {"child": "idle-waiting", "messages": [], "request_id": 0}
        state["feed"] = feed
    feed["messages"] = _as_message_list(feed.get("messages"))
    feed["request_id"] = int(feed.get("request_id") or 0)
    if feed.get("child") not in ("idle-waiting", "chat-assistant"):
        feed["child"] = "idle-waiting"
    return {"state": state, "effects": effects}


def _as_list(value: Any) -> list[Any]:
    if value is None:
        return []
    if isinstance(value, list):
        return value
    if isinstance(value, dict):
        if not value:
            return []
        if all(isinstance(key, int) for key in value):
            return [value[index] for index in range(1, max(value) + 1)]
    raise CoreError("Lua effects must be a list")


def _as_message_list(value: Any) -> list[Any]:
    if value is None or value == {}:
        return []
    if isinstance(value, list):
        return value
    if isinstance(value, dict) and all(isinstance(key, int) for key in value):
        return [value[index] for index in range(1, max(value) + 1)]
    raise CoreError("Conversation messages must be a list")

File: packages/policy-runtime/src/runtime.ts
export type Json =
  | null
  | boolean
  | number
  | string
  | Json[]
  | { [key: string]: Json };

export type PolicyEvent = {
  type: string;
  [key: string]: unknown;
};

export type PolicyEffect = {
  type: string;
  [key: string]: unknown;
};

export type PolicyResult = {
  state: unknown;
  effects?: PolicyEffect[];
  payload?: Record<string, unknown>;
};

export class PolicyError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "PolicyError";
  }
}

export type Helpers = {
  delegate: (name: string, state: unknown, event: unknown) => PolicyResult;
};

export type Transition = (
  state: unknown,
  event: PolicyEvent,
  helpers: Helpers,
) => PolicyResult;

export type PolicyContract = {
  scripts: Record<string, Transition>;
  effects?: Iterable<string>;
  maxDelegateDepth?: number;
  normalize?: (result: PolicyResult) => PolicyResult;
};

export type PolicyRuntime = {
  run(script: string, state: unknown, event: unknown): PolicyResult;
};

const DEFAULT_MAX_DEPTH = 8;

export function createRuntime(contract: PolicyContract): PolicyRuntime {
  const allowedEffects = contract.effects ? new Set(contract.effects) : null;
  const maxDepth = contract.maxDelegateDepth ?? DEFAULT_MAX_DEPTH;

  function runScript(
    script: string,
    state: unknown,
    event: unknown,
    depth: number,
  ): PolicyResult {
    if (!Object.hasOwn(contract.scripts, script)) {
      throw new PolicyError(`unknown policy script: ${script}`);
    }
    if (depth > maxDepth) {
      throw new PolicyError(`policy delegate depth exceeded: ${script}`);
    }

    const helpers: Helpers = {
      delegate(name: string, childState: unknown, childEvent: unknown) {
        const nextState =
          childState !== null && typeof childState === "object" && !Array.isArray(childState)
            ? childState
            : {};
        const nextEvent =
          childEvent !== null && typeof childEvent === "object" && !Array.isArray(childEvent)
            ? childEvent
            : {};
        return runScript(String(name), nextState, nextEvent, depth + 1);
      },
    };

    const raw = contract.scripts[script](
      state,
      asEvent(event),
      helpers,
    );
    if (raw === null || typeof raw !== "object" || Array.isArray(raw)) {
      throw new PolicyError(`policy transition must return an object: ${script}`);
    }

    const result: PolicyResult = {
      state: raw.state,
      effects: asEffectList(raw.effects),
      payload:
        raw.payload !== undefined &&
        raw.payload !== null &&
        typeof raw.payload === "object" &&
        !Array.isArray(raw.payload)
          ? raw.payload
          : undefined,
    };

    if (allowedEffects) {
      for (const effect of result.effects ?? []) {
        if (!allowedEffects.has(effect.type)) {
          throw new PolicyError(`unsupported client effect: ${JSON.stringify(effect)}`);
        }
      }
    }

    return contract.normalize ? contract.normalize(result) : result;
  }

  return {
    run(script: string, state: unknown, event: unknown) {
      return runScript(script, state, event, 0);
    },
  };
}

function asEvent(event: unknown): PolicyEvent {
  if (event !== null && typeof event === "object" && !Array.isArray(event)) {
    const type = (event as { type?: unknown }).type;
    return { ...(event as Record<string, unknown>), type: String(type ?? "") };
  }
  return { type: "" };
}

function asEffectList(value: unknown): PolicyEffect[] {
  if (value === undefined || value === null) {
    return [];
  }
  if (Array.isArray(value)) {
    return value.map((item) => {
      if (item === null || typeof item !== "object" || Array.isArray(item)) {
        throw new PolicyError(`unsupported client effect: ${item}`);
      }
      return item as PolicyEffect;
    });
  }
  throw new PolicyError("Policy effects must be a list");
}
